README.md

# Joulescope Server

Welcome to the pyjoulescope_server package!  This package enables Joulescopes
and [pyjoulescope](https://github.com/jetperch/pyjoulescope) to communicate 
over a network and interface with other programming languages using sockets.


## Quick start

Install Python for your platform.  See the 
[installation instructions](https://joulescope.readthedocs.io/en/latest/user/install.html)
for pyjoulescope for details.  Then:

    pip3 install -U joulescope_server

You should then be able to run the server:

    joulescope_server
    
If you would prefer to run directly from the clone git repo:

    python3 -m joulescope_server

To demonstrate the server, you can run the example client from
another terminal:

    python3 -m joulescope_server client


## License

All pyjoulescope_server code is released under the permissive Apache 2.0 license.
See the [License File](LICENSE.txt) for details.


CHANGELOG.md

# CHANGELOG

This file contains the list of changes made to pyjoulescope_ui.


## 0.1.5

2022 Dec 9

* Unregister statistics callback on device close #4.
* Stopped streaming and statistics on socket close.


## 0.1.4

2022 Nov 11

* Updated imports for joulescope 1.x.
* Fixed asyncio usage for python 3.8+.
* Updated README.


## 0.1.3

2022 Nov 8

* Improved error handling.


## 0.1.2

2021 Sep 27

* Fixed parameter_set.
* Properly handle CTRL-C.
* Default test client to  current range "auto".
* Log all send and receive messages.


## 0.1.1

2020 June 14

*   Initial development, now with streaming data.


joulescope_server/test/test_framer.py
# Copyright 2020 Jetperch LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


from joulescope_server import framer
import unittest


class TestFramer(unittest.TestCase):

    def test_pad(self):
        a = [0, 7, 6, 5, 4, 3, 2, 1, 0, 7, 6, 5, 4, 3, 2, 1, 0, 7, 6, 5, 4, 3, 2, 1, 0]
        for length, k in enumerate(a):
            self.assertEqual(k, framer.payload_length_to_pad_length(length))

    def test_header(self):
        tag1, len1 = b'ABN', 8
        header = framer.header_pack(tag1, len1)
        self.assertEqual(b'ABN\x00\x08\x00\x00\x00', header)
        tag2, len2 = framer.header_unpack(header)
        self.assertEqual(tag1, tag2)
        self.assertEqual(len1, len2)

    def test_pack_no_pad(self):
        payload = b'12345678'
        data = framer.pack(framer.TAG_ABN, b'12345678')
        self.assertEqual(b'ABN\x00\x08\x00\x00\x0012345678', data)
        tag, length = framer.header_unpack(data)
        self.assertEqual(b'ABN', tag)
        self.assertEqual(8, len(payload))

    def test_pack_with_pad(self):
        data = framer.pack(framer.TAG_ABN, b'hello')
        self.assertEqual(b'ABN\x00\x05\x00\x00\x00hello\x00\x00\x00', data)

    def test_unpack_multiple(self):
        d1 = framer.pack(framer.TAG_ABN, b'hello 1')
        d2 = framer.pack(framer.TAG_ABN, b'hello there 2')
        d = b''.join([d1, d2])
        t1, p1, d = framer.unpack(d)
        t2, p2, d = framer.unpack(d)
        self.assertEqual(t1, framer.TAG_ABN)
        self.assertEqual(t2, framer.TAG_ABN)
        self.assertEqual(p1, b'hello 1')
        self.assertEqual(p2, b'hello there 2')
        self.assertEqual(0, len(d))


joulescope_server/client.py
# Copyright 2020 Jetperch LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


import asyncio
from joulescope_server import framer, PORT
import numpy as np
import logging


class Client:

    def __init__(self, reader, writer):
        self._id = 0
        self._reader = reader
        self._writer = writer
        self._rsp_queue = asyncio.Queue()
        self._log = logging.getLogger(__name__)

    def _on_streaming_data(self, data):
        self._log.debug('streaming data')
        t1, p1, data = framer.unpack(data)
        if t1 != framer.TAG_AJS:
            self._log.warning('streaming data does not start with AJS')
            return
        info = framer.ajs_receive(p1)
        for field in info['data']['fields']:
            t, p, data = framer.unpack(data)
            # p contains the binary data for field
            if field in ['current', 'voltage', 'power']:
                # p for current, voltage, power is float32
                v = np.frombuffer(p, dtype=np.float32)
                print(f'{field}: {len(v)}')

    async def receiver(self):
        while True:
            try:
                tag, payload = await framer.receive(self._reader)
                if tag == framer.TAG_ABN:
                    self._on_streaming_data(payload)
                elif tag == framer.TAG_AJS:
                    msg = framer.ajs_receive(payload)
                    phase = msg.get('phase')
                    if phase == 'rsp':
                        # self._log.info('rsp: %s', msg)
                        await self._rsp_queue.put(msg)
                    elif phase == 'async':
                        self._log.info('async: %s', msg)
                        pass  # todo - handle async messages
                else:
                    self._log.warning('invalid phase %s', phase)
            except asyncio.IncompleteReadError:
                break
            except Exception:
                self._log.exception('handle_client error')
                return

    async def transact(self, type_, **kwargs):
        id_, self._id = self._id, self._id + 1
        req = {'id': id_, 'type': type_, 'phase': 'req'}
        req.update(**kwargs)
        self._log.info('req: %s', req)
        self._writer.write(framer.tpack(req))
        while True:
            rsp = await self._rsp_queue.get()
            if rsp['id'] != id_:
                self._log.warning('response id mismatch: expected %s, received %s',
                                  id_, rsp['id'])
                continue
            if rsp['status'] != 200:
                raise RuntimeError(rsp['status_msg'])
            break
        self._log.info('rsp: %s', rsp)
        return rsp


async def client_test_01():
    reader, writer = await asyncio.open_connection('127.0.0.1', PORT)
    client = Client(reader, writer)
    receiver_task = asyncio.create_task(client.receiver())

    rsp = await client.transact('hello')
    rsp = await client.transact('scan', config='auto')
    if not len(rsp['data']):
        print('Device not found')
        writer.close()
        return
    device = rsp['data'][0]

    rsp = await client.transact('open', device=device)
    rsp = await client.transact('parameters', device=device)
    rsp = await client.transact('info', device=device)
    rsp = await client.transact('parameter_get', device=device, data={'name': 'i_range'})
    rsp = await client.transact('parameter_set', device=device, data={'name': 'i_range', 'value': 'auto'})
    rsp = await client.transact('start', device=device, data={'fields': ['current', 'voltage']})
    await asyncio.sleep(3.0)  # receive statistics and streaming data
    rsp = await client.transact('stop', device=device)
    rsp = await client.transact('close', device=device)
    print('Close the socket')
    writer.close()
    await receiver_task


async def client_test_02():
    reader, writer = await asyncio.open_connection('127.0.0.1', PORT)
    client = Client(reader, writer)
    receiver_task = asyncio.create_task(client.receiver())
    rsp = await client.transact('hello')
    await asyncio.sleep(10.0)
    print('Close the socket')
    writer.close()
    await receiver_task


def run():
    loop = asyncio.get_event_loop()
    loop.run_until_complete(client_test_01())
    loop.close()


if __name__ == '__main__':
    logging.basicConfig(level=logging.INFO, format="%(levelname)s:%(asctime)s:%(filename)s:%(lineno)d:%(name)s:%(message)s")
    run()


joulescope_server/framer.py
# Copyright 2020 Jetperch LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


import struct
import json


TAG_AJS = b'AJS'    # json payload
TAG_ABN = b'ABN'    # binary payload
HEADER_LENGTH = 8   # tag + length


def payload_length_to_pad_length(payload_length):
    p = payload_length % 8
    if p > 0:
        p = 8 - p
    return p


def header_pack(tag, length):
    if len(tag) != 3:
        raise ValueError(f'invalid tag length {len(tag)}')
    if not isinstance(tag, bytes):
        raise ValueError(f'invalid tag type {type(tag)}')
    length = int(length)
    if length < 0 or length >= 2**24:
        raise ValueError(f'length out of range: {length}')
    length_bytes = struct.pack('<I', length)
    return tag + b'\x00' + length_bytes


def header_unpack(header):
    if len(header) < HEADER_LENGTH:
        raise ValueError('header is too short')
    tag = header[0:3]
    length_bytes = header[4:8]
    length = struct.unpack('<I', length_bytes)[0]
    return tag, length


def pack(tag, payload):
    if not isinstance(payload, bytes):
        raise ValueError(f'invalid payload type {type(payload)}')
    header_pack(tag, len(payload))
    length = len(payload)
    length_bytes = struct.pack('<I', len(payload))
    pad = payload_length_to_pad_length(length)
    pad_bytes = bytes([0] * pad)
    return tag + b'\x00' + length_bytes + payload + pad_bytes


def tpack(payload):
    if isinstance(payload, dict):
        return pack(TAG_AJS, json.dumps(payload).encode('utf-8'))
    else:
        return pack(TAG_ABN, payload)


def unpack(data):
    """Unpack one message.

    :param data: The bytes data to unpack.
    :return: (tag, payload, data_remaining).
    """
    tag, length = header_unpack(data)
    pad = payload_length_to_pad_length(length)
    total_length = HEADER_LENGTH + length + pad
    if len(data) < total_length:
        raise ValueError('insufficient data length')
    payload = data[HEADER_LENGTH:(HEADER_LENGTH + length)]
    return tag, payload, data[total_length:]


async def receive(reader):
    header = await reader.readexactly(HEADER_LENGTH)
    tag, length = header_unpack(header)
    pad = payload_length_to_pad_length(length)
    data = await reader.readexactly(length + pad)
    if not pad:
        payload = data
    else:
        payload = data[:-pad]
    return tag, payload


def ajs_receive(payload):
    return json.loads(payload.decode('utf-8'))


async def treceive(reader):
    tag, payload = await receive(reader)
    if tag == TAG_AJS:
        return ajs_receive(payload)
    elif tag == TAG_ABN:
        return payload
    else:
        raise ValueError('unsupported tag: tag=%s, payload=%s' % (tag, payload))


joulescope_server/__main__.py
# Copyright 2020 Jetperch LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


from joulescope_server.server import run as run_server
from joulescope_server.client import run as run_client
import argparse
import logging


def _run_server(p):
    return run_server()


def _run_client(p):
    return run_client()


def get_parser():
    """Run the Joulescope socket server or client."""
    parser = argparse.ArgumentParser(description='Joulescope socket server and client')
    parser.set_defaults(func=_run_server)

    s = parser.add_subparsers(title='command', help='sub-command help')

    p = s.add_parser('server', help='Run the server')
    p.set_defaults(func=_run_server)

    p = s.add_parser('client', help='Run the client')
    p.set_defaults(func=_run_client)

    return parser


def run():
    logging.basicConfig(level=logging.INFO, format="%(levelname)s:%(asctime)s:%(filename)s:%(lineno)d:%(name)s:%(message)s")
    p = get_parser()
    args = p.parse_args()
    return args.func(args)


if __name__ == '__main__':
    run()


joulescope_server/server.py
# Copyright 2020 Jetperch LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


"""
This server uses a simple tag-length-value (TLV) method to frame
each message.  The TLV format is:

3 bytes: tag.
1 byte: reserved for payload encoding.
4 bytes: payload length N, but current protocol max is 2**24.
N bytes: payload.
P bytes: 0 pad so next tag starts at multiple of 8 bytes.

The protocol currently defines two tags:

AJS: application-specific JSON format.
ABN: application-specific binary format

The AJS messages contain UTF-8, JSON-formatted data structures.
Each data structure has the following keys:

* type: The message type.
* phase: The message phase, one of [req, rsp, async].
  * req: Client request to the server
  * rsp: Server response to a client request
  * async: Server message generated asynchronously.
* id: The message id used to match rsp with req
* status: For rsp phase, 200 or error code.
* status_msg: For rsp phase, 'Success' or the error message.
* device: The name of the target device returned by "scan".
* data: The data, which is dependent upon type.


The available req/rsp message types are:

* hello: rsp data contains version information.
* scan: rsp data is the list of device names.
* open
* close
* info: data contains device info metadata.
* parameters: data contains dict of parameter name to metadata.
* parameter_get: data contains
  * 'name': the parameter name.
  * 'value': on response, contains the parameter value.
* parameter_set
  * 'name': the parameter name.
  * 'value': the new parameter value.
* start: data contains
    * fields: list containing any valid StreamBuffer.samples_get field
        [current, voltage, power,
         bits, current_range, current_lsb, voltage_lsb,
         raw, raw_current, raw_voltage]
    * sample_chunk: number of samples to send in each update.
        Defaults to 1/10 of a second.
* stop

The async message types are:

* device_notify - Device inserted or removed, issue "scan" for details.
* statistics
* event
* stop
"""


import asyncio
from joulescope_server import framer, PORT, __version__
import json
import logging
import joulescope
from joulescope import DeviceNotify
import weakref


def msg_status(msg, status_code, status_msg):
    msg['status'] = status_code
    msg['status_msg'] = status_msg


def _param_to_dict(p):
    return {
        'name': p.name,
        'default': p.default,
        'path': p.path,
        'options': p.options,
        'units': p.units,
        'brief': p.brief,
        'detail': p.detail,
        'flags': p.flags,
    }


class DeviceStreamManager:

    def __init__(self, parent, device_name, sample_chunk, fields):
        self._parent = weakref.ref(parent)
        self._device_name = device_name
        self._sample_chunk = sample_chunk
        self._fields = fields
        self._sample_id_last = None
        self._stream_buffer = None
        self._log = logging.getLogger(__name__ + '.device_name')

    def start(self, stream_buffer):
        self._stream_buffer = stream_buffer
        self._sample_id_last = stream_buffer.sample_id_range[-1]

    def _update(self, force=False):
        # called from USB thread, must resynchronize
        if self._stream_buffer is None:
            return
        idx_start, idx_stop = self._stream_buffer.sample_id_range
        if idx_stop <= self._sample_id_last:
            return
        if idx_start < self._sample_id_last:
            idx_start = self._sample_id_last
        if not force and (idx_start + self._sample_chunk) < idx_stop:
            return
        msg = {
            'type': 'stream_data',
            'phase': 'bin',
            'device': self._device_name,
            'data': {
                'stream_buffer': self._stream_buffer,
                'sample_range': [idx_start, idx_stop],
                'fields': self._fields,
            },
        }
        self._parent()._async_queue_put_threadsafe(msg)
        self._sample_id_last = idx_stop

    def stop(self):
        self._update(True)
        self._stream_buffer = None

    def stream_notify(self, stream_buffer):
        if stream_buffer != self._stream_buffer:
            self._log.warning('stream buffer mismatch')
            self._stream_buffer = stream_buffer
        self._update()

    def close(self):
        pass


class ClientManager:
    _instances = set()  # weakreaf to instances

    def __init__(self, reader, writer):
        self._devices = []
        self._device_streaming = {}  # Map(device_name: str, DeviceStreamManager)
        self._statistics_fn_by_device = {}  # Map(device_name: str, callable)
        self._reader = reader
        self._writer = writer
        self._async_queue = asyncio.Queue()
        self._async_task = asyncio.create_task(self._async_task())
        self._loop = asyncio.get_event_loop()
        self._log = logging.getLogger(__name__)
        ClientManager._instances.add(weakref.ref(self))

    @classmethod
    def cls_handle_device_notify(cls, inserted, info):
        invalid = set()
        for instance in ClientManager._instances:
            c = instance()
            if c is None:
                invalid.add(instance)
            else:
                c.handle_device_notify()
        cls._instances -= invalid

    def handle_device_notify(self):
        self._log.info('_handle_device_notify()')
        msg = {
            'type': 'device_notify',
            'phase': 'async',
        }
        self._async_queue_put_threadsafe(msg)

    def device_get_by_name(self, device_name):
        for d in self._devices:
            if str(d) == device_name:
                return d
        raise ValueError('device not found')

    def device_get(self, msg):
        if 'device' not in msg:
            raise ValueError('device not specified')
        return self.device_get_by_name(msg['device'])

    async def _on_hello(self, msg):
        msg['data'] = {
            'protocol_version': 1,
            'server_version': __version__,
            'joulescope_version': joulescope.__version__,
        }
        return msg

    def _async_queue_put_threadsafe(self, msg):
        self._loop.call_soon_threadsafe(self._async_queue.put_nowait, msg)

    async def _on_stream_data(self, msg):
        device_name = msg['device']
        if device_name not in self._device_streaming:
            self._log.info('_on_stream_data, but device %s not streaming', device_name)
            return
        data = msg['data']
        stream_buffer = data.pop('stream_buffer')
        start, stop = data['sample_range']
        d = stream_buffer.samples_get(start, stop, data['fields'])
        msg['data']['time'] = d['time']
        fields = []
        for field in data['fields']:
            k = d['signals'][field]
            fields.append([field, k['units']])
        payload = [framer.tpack(msg)]
        for field in data['fields']:
            v = d['signals'][field]['value']
            payload.append(framer.tpack(v.tobytes()))
        return b''.join(payload)

    async def _async_task(self):
        self._log.info('_async_task start')
        while True:
            try:
                msg = await self._async_queue.get()
                msg_type = msg['type']
                if msg_type == 'close':
                    break
                elif msg_type == 'stream_data':
                    msg = await self._on_stream_data(msg)
                else:
                    msg['phase'] = 'async'
                    msg['status'] = 200
                    msg['status_msg'] = 'Success'
                if msg is not None:
                    self._writer.write(framer.tpack(msg))
            except Exception:
                self._log.exception('_async_task')

        # Stop streaming
        while len(self._device_streaming):
            device_name, stream_process = self._device_streaming.popitem()
            d = self.device_get_by_name(device_name)
            d.stop()
            d.stream_process_unregister(stream_process)

        # Stop statistics
        while len(self._statistics_fn_by_device):
            device_name, statistics_fn = self._statistics_fn_by_device.popitem()
            d = self.device_get_by_name(device_name)
            d.statistics_callback_unregister(statistics_fn)

        self._log.info('_async_task done')

    async def _on_scan(self, msg):
        config = msg.get('config', 'auto')
        self._devices, _, _ = joulescope.scan_for_changes('joulescope', self._devices, config=config)
        msg['data'] = [str(d) for d in self._devices]
        return msg

    def _statistics_fn(self, device_name, data):
        self._log.info('_statistics_fn %s', device_name)
        msg = {
            'type': 'statistics',
            'phase': 'async',
            'device': device_name,
            'data': data,
        }
        self._async_queue_put_threadsafe(msg)

    def _event_fn(self, device_name, event, message):
        self._log.info('_event_fn %s', device_name)
        msg = {
            'type': 'event',
            'phase': 'async',
            'device': device_name,
            'data': {
                'event': event,
                'message': message,
            },
        }
        self._async_queue_put_threadsafe(msg)

    async def _on_open(self, msg):
        d = self.device_get(msg)
        dname = str(d)
        statistics_fn = lambda data: self._statistics_fn(dname, data)
        self._statistics_fn_by_device[dname] = statistics_fn
        d.statistics_callback_register(statistics_fn)
        d.open(lambda event, message: self._event_fn(dname, event, message))
        return msg

    def _statistics_unregister(self, device):
        dname = str(device)
        statistics_fn = self._statistics_fn_by_device.pop(dname, None)
        if statistics_fn is not None:
            device.statistics_callback_unregister(statistics_fn)

    async def _on_close(self, msg):
        d = self.device_get(msg)
        self._statistics_unregister(d)
        d.close()
        return msg

    async def _on_parameters(self, msg):
        d = self.device_get(msg)
        params = d.parameters()
        msg['data'] = {p.name: _param_to_dict(p) for p in params}
        return msg

    async def _on_parameter_get(self, msg):
        d = self.device_get(msg)
        msg['data'] = d.parameter_get(msg['data']['name'])
        return msg

    async def _on_parameter_set(self, msg):
        d = self.device_get(msg)
        d.parameter_set(msg['data']['name'], msg['data']['value'])
        return msg

    async def _on_info(self, msg):
        d = self.device_get(msg)
        msg['data'] = d.info()
        return msg

    def _stop_fn(self, device_name, event, message):
        self._log.info('_stop_fn %s', device_name)
        msg = {
            'type': 'stop',
            'phase': 'async',
            'device': device_name,
            'data': {
                'event': event,
                'message': message,
            },
        }
        self._async_queue_put_threadsafe(msg)

    async def _on_start(self, msg):
        d = self.device_get(msg)
        dname = str(d)
        data = msg.get('data', {})
        fields = data.get('fields', ['current', 'voltage'])
        sample_chunk = data.get('sample_chunk', d.output_sampling_frequency / 10)
        stream_process = DeviceStreamManager(self, dname, sample_chunk, fields)
        self._device_streaming[dname] = stream_process
        d.stream_process_register(stream_process)
        msg['data'] = d.start(lambda event, message: self._stop_fn(dname, event, message))
        return msg

    async def _on_stop(self, msg):
        d = self.device_get(msg)
        msg['data'] = d.stop()
        stream_process = self._device_streaming.pop(str(d), None)
        d.stream_process_unregister(stream_process)
        return msg

    async def _on_unknown(self, msg):
        msg_status(msg, 404, 'Type not found')
        return msg

    async def run(self):
        self._log.info('ClientManager.run start')
        while True:
            try:
                msg = await framer.treceive(self._reader)
                self._log.info(f'recv: {json.dumps(msg)}')
                print()
                phase = msg.get('phase', 'req')
                if phase != 'req':
                    msg['error'] = f'Invalid phase: {phase}'
                    self._writer.write(msg)
                    continue
                msg['phase'] = 'rsp'
                msg_status(msg, 200, 'Success')
                type_ = msg.get('type', None)
                # automatically bind type to method by name
                method_name = f'_on_{type_}'
                fn = getattr(self, method_name, self._on_unknown)
                try:
                    msg = await fn(msg)
                except Exception as ex:
                    self._log.exception(f'fn type {type_}')
                    msg_status(msg, 500, 'Error')
                if msg is not None:
                    self._log.info(f'send: {json.dumps(msg)}')
                    self._writer.write(framer.tpack(msg))
            except asyncio.IncompleteReadError:
                self._log.info('Client closed socket')
                break
            except Exception:
                self._log.exception('handle_client error')
                break
        try:
            await self._writer.drain()
            self._writer.close()
        except Exception:
            self._log.warning('Writer did not close gracefully')
        try:
            await self._async_queue.put({'type': 'close'})
            await self._async_task
        except Exception:
            self._log.exception('Could not close client task gracefully')
        self._log.info('ClientManager.run done')


async def handle_client(reader, writer):
    mgr = ClientManager(reader, writer)
    await mgr.run()


async def run_async():
    server = await asyncio.start_server(handle_client, '127.0.0.1', PORT)
    print('Serving on {}'.format(server.sockets[0].getsockname()))
    await server.serve_forever()


def run():
    device_notify = DeviceNotify(ClientManager.cls_handle_device_notify)
    try:
        asyncio.run(run_async())
    except KeyboardInterrupt:
        pass
    for device in joulescope.scan():
        try:
            device.close()
        except Exception:
            pass
    device_notify.close()


if __name__ == '__main__':
    logging.basicConfig()
    run()
